### PR TITLE
IFU 2.12 JAX merge conflict resolution

### PR DESCRIPTION
**Pull request by Micky774**
# Description

This PR focuses on the resolutions to merge conflicts located in the PyTorch integration, as introduced as part of the main IFU branch. This merge resolution is performed in separate PRs to limit the scope of each PR and simplify the review process, while preserving commit history and conversation readability.

This is one of three PRs which target the common components of TE, as well as the JAX and PyTorch integrations. Once all three PRs are resolved and merged into the IFU branch, a new IFU PR will be created with a focus on ensuring a successful total build as well as CI validation. In the meanwhile, the PRs will be partially validated via local testing.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

The changelog is generated with assistance from Claude.

  1. `tests/jax/test_fused_attn.py` — four conflict resolutions:
    - Kept ROCm's new-style RNG key handling (jax.random.key vs PRNGKey) while adopting upstream's 6-way split (added softmax_key)
    - Kept ROCm's RNG partition spec logic while integrating upstream's new softmax offset sharding
    - Fixed test param ID for cross-attention from seqlen 512 to 2048 while keeping upstream's `RAGGED_KV_PACKED` layout addition
    - Added `pytest.skip` for `test_context_parallel_allgather_striped_attn` on ROCm — THD + ALL_GATHER + Striped attention is not yet supported
  2. `tests/jax/utils.py` — resolved conflict keeping both ROCm's MXFP8 GEMM support-checking functions and upstream's new `is_devices_equal` utility
  3. `jax/csrc/extensions/attention.cpp` — fixed platform guard structure for fused attention aux tensors: ROCm enters the arbitrary-seqlen code path unconditionally (both aotriton and CK backends need it), while CUDA only enters for
  `NVTE_F16_arbitrary_seqlen`. Also integrated upstream's new softmax_offset aux tensor with proper `#ifndef USE_ROCM` guard
  4. `jax/csrc/extensions/gemm.cpp` — resolved conflict merging ROCm's NVFP4 guard ("ROCm TE does not support NVFP4 yet") with upstream's new `set_with_gemm_swizzled_scales` API. Adopted upstream's code structure while preserving the ROCm NVFP4 error

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_distributed_fused_attn.py [hunk 1]
```diff
@@ -194,9 +194,7 @@ def test_self_attn(
             pytest.param(AttnBiasType.PRE_SCALE_BIAS, BiasShape._1HSS, id="PRE_SCALE_BIAS-1HSS"),
         ],
     )
-<<<<<<< HEAD
     @pytest.mark.skipif(version.parse(jax.__version__) < version.parse("0.5.0"), reason="shardy sharding requires JAX 0.5.0")
```
Comment (ipanfilo) [inline tests/jax/test_distributed_fused_attn.py]: This condition can be removed. JAX>=0.5.0 is currently required

File: tests/jax/test_fused_attn.py [hunk 2]
```diff
@@ -820,10 +804,15 @@ def to_dp_shardings(x):
         self.softmax_offset_pspec = PartitionSpec(None, head_resource, None, None)
         self.softmax_offset_sharding = NamedSharding(self.mesh, self.softmax_offset_pspec)
 
-        self.dropout_rng_pspec = PartitionSpec(
-            None,
-        )
->>>>>>> 99df88
+        # New-style RNG fix is only applied for AMD GPUs
+        self.dropout_rng_pspec = PartitionSpec(None,)
```
Comment (ipanfilo) [inline tests/jax/test_fused_attn.py]: Keep upstream variant of this call then (deleted lines 823-825)

File: tests/jax/test_fused_attn.py [hunk 3]
```diff
@@ -1160,12 +1149,8 @@ def check_dqkv(primitive, reference, pad, idx):
             64,
             64,
             jnp.bfloat16,
-<<<<<<< HEAD
-            id="2-2048-1024-12-12-64-64-BF16-CROSS",
-=======
             QKVLayout.THD_T2HD,
-            id="2-512-1024-12-12-64-64-BF16-CROSS-RAGGED_KV_PACKED",
->>>>>>> 99df88
+            id="2-2048-1024-12-12-64-64-BF16-CROSS-RAGGED_KV_PACKED",
```
Comment (ipanfilo) [inline tests/jax/test_fused_attn.py]: It is "mid data size + bf16" test config, why is it 2-2048, not 2-512 like in upstream?

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: tests/jax/test_fused_attn.py (hunk 3 above)
Comment (Micky774) [reply tests/jax/test_fused_attn.py]: ~@wangye805 the use of `2048` instead of `512` came from your initial commit, is there a particular reason for the divergence or was it just missed initially?~

Edit: actually it looks like it was just a mistaken diff match on git's part, I'll update it

File: tests/jax/test_fused_attn.py (hunk 2 above)
Comment (Micky774) [reply tests/jax/test_fused_attn.py]: Updated

File: tests/jax/test_distributed_fused_attn.py (hunk 1 above)
Comment (Micky774) [reply tests/jax/test_distributed_fused_attn.py]: Updated

**Review by ipanfilo — APPROVED** [review]

**Comment by Micky774** [comment]
Closing as completed -- no merge needed
